### PR TITLE
Start adding documentation for metaprogramming

### PR DESCRIPTION
**Pull request by bollu**
I wish to write down the examples of using LEAN4 macros that I feel are useful to get up and running. This is part 1 which parses the simplest non-trivial grammar I can think of. 

I know very little of how the macro system works internally, so corrections and elaborations would be very useful!

### CODE REVIEW

**Comment by leodemoura** [comment]
Thanks a bunch for writing it down. I have only a few comments.

File: doc/metaprogramming.md [hunk 1]
```diff
@@ -1 +1,139 @@
 # Metaprogramming
+
+Macros are a language feature that allows writing code that writes other code (metaprogramming).
+
+In LEAN4, macros are used pervasively. So much so that core language features such as `do` notation
```
Comment (leodemoura) [inline doc/metaprogramming.md:5]: We use "Lean 4" is the rest of the documentation.

File: doc/metaprogramming.md [hunk 2]
```diff
@@ -1 +1,139 @@
...
+is implemented via macros! As a language user, macros are useful to easily
+embed domain-specific languages and to generate code at compile-time, to name a few uses.
+
+
+#### Balanced Parentheses as an Embedded Domain Specific Language
+
+Let's look at how to use macros to extend the LEAN parser and embed a language for building _balanced parentheses_.
```
Comment (leodemoura) [inline doc/metaprogramming.md:12]: "LEAN" => "Lean"

File: doc/metaprogramming.md [hunk 3]
```diff
@@ -1 +1,139 @@
...
+We begin by defining an inductive data type of the grammar we wish to parse:
+
+```lean
+inductive Dyck: Type :=
+   | Round : Dyck -> Dyck -- ( <inner> )
+   | Flower : Dyck -> Dyck -- { <inner> }
+   | End : Dyck
+```
+
+We begin by declaring a _syntax category_ using the `declare_syntax_cat <category>` directive.
```
Comment (leodemoura) [inline doc/metaprogramming.md:31]: "directive" => "command"

File: doc/metaprogramming.md [hunk 4]
```diff
@@ -1 +1,139 @@
...
+The above means that the string "End" lives in syntax category `brack`.
+
+Similarly, we declare the rules `"(" Dyck ")"` and `"{" Dyck "}"` using the rules:
+
+```lean
+syntax "(" brack ")" : brack
+syntax "{" brack "}" : brack
+```
+
+Finally, we need a way to build _LEAN4 terms_ from this grammar -- that is, we must translate out of this
```
Comment (leodemoura) [inline doc/metaprogramming.md:53]: "LEAN4" => "Lean 4"

**Comment by Kha** [comment]
Note that this will not be visible unless added to SUMMARY.md. how should this be differentiated from syntax.md? Should we have a new folder/section for tutorials?

**Comment by leodemoura** [comment]
> Should we have a new folder/section for tutorials?

I think this is a great idea.

File: doc/metaprogramming.md [hunk 5]
```diff
@@ -1 +1,139 @@
...
+#### Balanced Parentheses as an Embedded Domain Specific Language
+
+Let's look at how to use macros to extend the Lean 4 parser and embed a language for building _balanced parentheses_.
+This language accepts strings given by the [BNF grammar](https://en.wikipedia.org/wiki/Backus%E2%80%93Naur_form)
+
+```
+Dyck := 
+  "(" Dyck ")" 
+  | "{" Dyck '}'
+  | "End"
```
Comment (Kha) [inline doc/metaprogramming.md:19]: This token looks like a hack. How can we make the grammar more conventional? We could use `(brack)?`, but that is not as nice in BNF. An epsilon production would be more common there, but we don't have a direct encoding of that as a `ParserDescr` yet. Should we?

File: doc/metaprogramming.md [hunk 6]
```diff
@@ -1 +1,139 @@
...
+```
+
+We begin by defining an inductive data type of the grammar we wish to parse:
+
+```lean
+inductive Dyck: Type :=
+   | Round : Dyck -> Dyck -- ( <inner> )
+   | Flower : Dyck -> Dyck -- { <inner> }
```
Comment (Kha) [inline doc/metaprogramming.md:27]: I don't think I've ever heard of flower brackets :) . Sounds very poetic though. Is it a common name?

File: doc/metaprogramming.md (hunk 5 above)
Comment (bollu) [reply doc/metaprogramming.md:19]: Having a `?` would be very useful!

File: doc/metaprogramming.md (hunk 6 above)
Comment (bollu) [reply doc/metaprogramming.md:27]: That's what I was taught as a child. It appears to be a colloquialism :) I'll be happy to change it to curly brackets if you'd like!

File: doc/metaprogramming.md (hunk 5 above)
Comment (Kha) [reply doc/metaprogramming.md:19]: We could also modify the `syntax` command to accept an empty sequence of parsers, but an explicit epsilon parser might be the better idea

File: doc/metaprogramming.md (hunk 6 above)
Comment (Kha) [reply doc/metaprogramming.md:27]: Yeah, feel free to change this file in the new PR so we have both files side by side

**Pull request merged**